underfs/b2/src/main/java/alluxio/underfs/b2/B2UnderFileSystem.java

```diff
@@ -0,0 +1,374 @@
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.UnderFileSystemConfiguration;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.CommonUtils;
+import alluxio.util.UnderFileSystemUtils;
+import alluxio.util.io.PathUtils;
+
+import com.backblaze.b2.client.B2ListFilesIterable;
+import com.backblaze.b2.client.B2StorageClient;
+import com.backblaze.b2.client.B2StorageClientFactory;
+import com.backblaze.b2.client.contentSources.B2ContentSource;
+import com.backblaze.b2.client.contentSources.B2ContentTypes;
+import com.backblaze.b2.client.contentSources.B2FileContentSource;
+import com.backblaze.b2.client.exceptions.B2Exception;
+import com.backblaze.b2.client.structures.B2AccountAuthorization;
+import com.backblaze.b2.client.structures.B2Allowed;
+import com.backblaze.b2.client.structures.B2Bucket;
+import com.backblaze.b2.client.structures.B2CopyFileRequest;
+import com.backblaze.b2.client.structures.B2FileVersion;
+import com.backblaze.b2.client.structures.B2ListBucketsRequest;
+import com.backblaze.b2.client.structures.B2ListFileNamesRequest;
+import com.backblaze.b2.client.structures.B2ListFileVersionsRequest;
+import com.backblaze.b2.client.structures.B2UploadFileRequest;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ServiceException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * B2 {@link UnderFileSystem} implementation based on the b2-sdk-core library.
+ */
+@ThreadSafe
+public class B2UnderFileSystem extends ObjectUnderFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(B2UnderFileSystem.class);
+
+  public static final String B2_SCHEME = "b2://";
+  private static final String FOLDER_SUFFIX = "/";
+
+  private final B2StorageClient mB2StorageClient;
+  private final String mAccountId;
+  private final String mBucketName;
+  private final String mBucketId;
+  private final String mFolderIndicator;
+  private final short mBucketMode;
+
+  protected B2UnderFileSystem(AlluxioURI alluxioURI, UnderFileSystemConfiguration conf,
+      B2StorageClient b2StorageClient, String bucketId, String bucketName, short bucketMode,
+      String accountId) {
+    super(alluxioURI, conf);
+    requireNonNull(alluxioURI, "alluxioURI is null");
+    requireNonNull(conf, "conf is null");
+    requireNonNull(b2StorageClient, "b2StorageClient is null");
+    requireNonNull(bucketId, "bucketId is null");
+    requireNonNull(bucketName, "bucketName is null");
+    requireNonNull(bucketMode, "bucketMode is null");
+    requireNonNull(accountId, "accountId is null");
+
+    LOG.debug("alluxioURI: {}", alluxioURI);
```

Comment:
If we want to keep this debug logging, can we combine it to a single log line that prints all of these parameters?